File: 295-find-median-from-data-stream/find-median-from-data-stream.cpp
class MedianFinder {
public:
   priority_queue<int> maxHeap;
   priority_queue<int,vector<int>, greater<int> > minHeap;
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        minHeap.push(num);
        int k = minHeap.top();
        minHeap.pop();
        maxHeap.push(k);
        if(maxHeap.size()> 1+ minHeap.size())
        {
            int t = maxHeap.top();
            maxHeap.pop();
            minHeap.push(t);
        }
        return;
    }
    
    double findMedian() {
        if(maxHeap.size()==minHeap.size()) return (double)(minHeap.top()+maxHeap.top())/2.00;
    else return (double)maxHeap.top();
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 295-find-median-from-data-stream/find-median-from-data-stream.java
class MedianFinder {
       PriorityQueue<Integer> minHeap;
       PriorityQueue<Integer> maxHeap;
    public MedianFinder() {
        minHeap = new PriorityQueue<>();//stores larger value
        maxHeap = new PriorityQueue<>(Comparator.reverseOrder());
        
    }
    
    public void addNum(int num) {
        minHeap.offer(num);
        maxHeap.offer(minHeap.poll());
        if(maxHeap.size()>1+minHeap.size())
        {
            minHeap.offer(maxHeap.poll());
        }
    }
    
    public double findMedian() {
        if(minHeap.size()==maxHeap.size()){
            return (double)(minHeap.peek()+maxHeap.peek())/2.00;
        }
        else return(double) maxHeap.peek();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */